src/plugins/AvatarPEPPlugin.ts

```diff
@@ -0,0 +1,245 @@
+            .then(function (result) {
+               if ($(result).attr('type') === 'result') {
+                  return [undefined];
+               } else {
+                  return [avatar];
+               }
+            });
+      } else {
+         let data = $build('data', { xmlns: UANS_DATA }).t(avatar.getData()).tree();
+
+         return connection
+            .getPEPService()
+            .publish(UANS_DATA, data, UANS_DATA)
+            .then((result: any) => {
+               if ($(result).attr('type') === 'result') {
+                  let metadata;
+                  let hash = Hash.SHA1FromBase64(avatar.getData());
+                  let i = new Image();
+                  let iheight = 0;
+                  let iwidth = 0;
+                  let size = FileHelper.getFileSizeFromBase64(avatar.getData());
+
+                  i.onload = function () {
+                     iheight = i.height;
+                     iwidth = i.width;
+                  };
+                  i.src = 'data:' + avatar.getType() + ';base64,' + avatar.getData();
+                  metadata = $build('metadata', { xmlns: UANS_METADATA })
+                     .c('info', { bytes: size, id: hash, height: iheight, width: iwidth, type: avatar.getType() })
+                     .tree();
+
+                  return connection
+                     .getPEPService()
+                     .publish(UANS_METADATA, metadata, UANS_METADATA)
+                     .then(function (result) {
+                        if ($(result).attr('type') === 'result') {
+                           return [undefined];
+                        } else {
+                           return [avatar];
+                        }
+                     });
+               } else {
+                  return [avatar];
+               }
+            });
+      }
+   };
+
+   private onMessageAvatarUpdate = stanza => {
+      let from = new JID($(stanza).attr('from'));
+      let metadata = $(stanza).find('metadata[xmlns="urn:xmpp:avatar:metadata"]');
+      let data = $(stanza).find('data[xmlns="urn:xmpp:avatar:data"]');
+
+      if (metadata.length > 0) {
+         let info = metadata.find('info');
+         let contact = this.pluginAPI.getContact(from);
+         if (!contact) {
+            this.pluginAPI.Log.warn('No contact found for', from);
+            return true;
+         }
+
+         if (info.length > 0) {
+            let hash = $(info).attr('id');
+
+            let storedHash = this.getStorage().getItem(from.bare);
+            if (storedHash === undefined || hash !== storedHash) {
+               let avatarUI = AvatarUI.get(contact);
+               this.getStorage().setItem(contact.getJid().bare, hash);
+               avatarUI.reload();
+            }
+         } else {
+            let avatarUI = AvatarUI.get(contact);
+            this.getStorage().setItem(contact.getJid().bare, '');
+            avatarUI.reload();
+         }
+      } else if (data.length > 0) {
```

Comment:
Since we subscribe only to metadata updates, this block should not be needed.

Reply:
Doesnt the server sends pep data to user when he has succesfully logged in ? Then it would be useful

Reply:
No. The server should not send any data without request or subscription.

Reply:
okay.. then i could delete it..